src/main/kotlin/com/doist/gradle/changelog/EntryValidator.kt

```diff
@@ -0,0 +1,54 @@
+package com.doist.gradle.changelog
+
+internal class EntryValidator(private val config: ValidatorConfig) {
+    private val localValidators = buildLocalValidators()
+    private val ignorePartRegex = config.ignorePart?.toRegex()
+
+    fun validate(entry: String): Result {
+        val entryWithoutIgnoredPart = ignorePartRegex?.let {
+            entry.replace(ignorePartRegex, "")
+        } ?: entry
+
+        val brokenRules = localValidators
+            .filterNot { it.isValid(entryWithoutIgnoredPart) }
+            .map { it.rule }
+
+        return when {
+            brokenRules.isEmpty() -> Result.Valid
+            else -> Result.Invalid(brokenRules)
+        }
+    }
+
+    private fun buildLocalValidators(): List<LocalValidator> {
+        val validators = mutableListOf<LocalValidator>()
+        if (config.maxLength > 0) {
+            validators.add(
+                LocalValidator({ it.length <= config.maxLength }, Rule.MaxLength(config.maxLength))
+            )
+        }
+        if (config.endsWithDot) {
+            validators.add(LocalValidator({ it.endsWith(".") }, Rule.EndsWithDot))
+        }
+        if (config.doesNotEndWithDot) {
+            validators.add(LocalValidator({ !it.endsWith(".") }, Rule.DoesNotEndWithDot))
+        }
+
+        return validators
+    }
+
+    sealed class Rule {
+        data class MaxLength(val limit: Int) : Rule()
+        object EndsWithDot : Rule()
+        object DoesNotEndWithDot : Rule()
+    }
+
+    sealed class Result {
+        object Valid : Result()
+        data class Invalid(val brokenRules: List<Rule>) : Result()
+    }
+
+    private data class LocalValidator(
+        val isValid: (String) -> Boolean,
+        val rule: Rule
+    )
```

Comment:
Currently each `Rule` uses the same function `isValid`. I suggest to have class `Rule` as

```kotlin

sealed class Rule {
    abstract fun isValid(entry: String): Boolean

    data class MaxLength(val limit: Int) : Rule() {
        override fun isValid(entry: String) = entry.length <= limit
    }
    object EndsWithDot : Rule() {
        override fun isValid(entry: String) = entry.endsWith(".")
    }
    object DoesNotEndWithDot : Rule() {
        override fun isValid(entry: String) = !entry.endsWith(".")
    }
}
```

and remove `LocalValidator`. 

In case of keeping `LocalValidator`, I'd suggest to pass lambda as the last parameter into constructor to make its usage more flexible with moving lambda outside of constructor. Additionally I'd simplify naming and rename it into `Validator` given that we don't have external validators :)

Reply:
That's a great suggestion! I think I've overcomplicated it 😅 We should simply allow creating any rule in the plugin configuration.